projects/seeker/agents/gpt2_seeker_modules.py

```diff
@@ -0,0 +1,553 @@
+        """
+        return FidModel.build_encoder(
+            opt, dictionary, encoder_class=IdentityLayer, **kwargs
+        )
+
+    @classmethod
+    def build_decoder(
+        cls,
+        opt: Opt,
+        *args,
+        embedding: Optional[torch.nn.Embedding] = None,
+        n_positions: Optional[int] = None,
+        decoder_class: Optional[Type] = None,
+        **kwargs,
+    ):
+        """
+        Override to use GPT2 as decoder.
+        """
+        return FidModel.build_decoder(
+            opt,
+            Gpt2Agent.dictionary_class()(opt),
+            decoder_class=ComboGPT2Decoder,
+            **kwargs,
+        )
+
+    def encoder(
+        self,
+        input: torch.LongTensor,
+        input_lengths: torch.LongTensor,
+        query_vec: torch.LongTensor,
+        input_turns_cnt: torch.LongTensor,
+        target_lengths: Optional[torch.LongTensor],
+        positions: Optional[torch.LongTensor] = None,
+        segments: Optional[torch.LongTensor] = None,
+    ) -> Tuple[
+        torch.Tensor,
+        torch.BoolTensor,
+        Optional[torch.LongTensor],
+        Optional[List[List[Document]]],
+        Optional[torch.Tensor],
+    ]:
+        """
+        Override FidModel.encoder to pack all the documents into one input example.
+
+        :param input:
+            2D [bsz, seqlen] input to the encoder
+        :param input_lengths:
+            1D [bsz] lengths of each input item
+        :param query_vec:
+            2D [bsz*n_turns, seqlen] input for the retriever
+        :param input_turns_cnt:
+            1D [bsz] number of dialogue turns for each input example
+        :param input_lengths:
+            1D [bsz] lengths of each target item (for each input item)
+
+        :return (encoder_out, encoder_mask, input_turns_cnt, top_docs, top_doc_scores):
+            encoder_out: *concatenated* encoded representations of context/document pairs
+            encoder_mask: new mask for enc_out
+            input_turns_cnt: pass along the input turns count for the decoder
+            top_docs: List of top Documents for each batch example
+            top_doc_scores: scores for each retrieved document.
+        """
+        enc_out, mask, input_turns_cnt, top_docs, top_doc_scores = RagModel.encoder(
+            self, input, input_lengths, query_vec, input_turns_cnt, positions, segments
+        )  # type: ignore
+        seq_len, n_docs = enc_out.size(1), enc_out.size(0) // input.size(0)
+
+        if input_turns_cnt is not None:
+            # Input Turns is a tensor of dim [bsz]
+            input = input.repeat_interleave(input_turns_cnt, dim=0)  # type: ignore
+        doc_starts = (enc_out == self.pad_idx).sum(dim=1)  # left padded
+        doc_lens = (seq_len - input_lengths.repeat_interleave(n_docs)) - doc_starts
+        # if no padding, docs are assumed to be min doc length long
+        doc_lens[doc_lens.le(0)] = self.min_doc_len
+        new_enc_out = enc_out.clone()
+        # BEFORE:
```

Comment:
Question: we had this before and after for FiD because we needed to encode the doc with the input. In this module that only does string operations, is it still a reason for doing this way other than using pre-existing methods for making masks etc.?

Reply:
well we're actually still doing operations on tensors, which is why it's a bit confusing. This mess of logic is a result of me trying to format a GPT2 decoder-style model into a FiD setup, rather than the other way around